modules/aws-codebuild-step/example/main.tf

```diff
@@ -0,0 +1,178 @@
+provider "aws" {
+  region = "us-east-1"
+}
+
+module "build" {
+  source = "../"
+
+  project_name    = "build-${var.environment}"
+  description     = "thing_that_build"
+  environment     = var.environment
+  pipeline_s3_arn = aws_s3_bucket.codepipeline_bucket.arn
+
+  buildspec_path = ".aws/dev/build.yaml"
+  build_image    = "node:14"
+  build_timeout  = "10"
+
+  environment_variable = [
+    {
+      name  = "ITS-DANGEROUS-TO-GO-ALONE-TAKE-THIS"
+      value = "Cat"
+    }
+  ]
+}
+
+
+module "deploy" {
+  source = "../"
+
+  project_name    = "website-deploy-${var.environment}"
+  description     = "thing_that_deploys_whatever_we_builded"
+  environment     = var.environment
+  pipeline_s3_arn = aws_s3_bucket.codepipeline_bucket.arn
+
+  buildspec_path = ".aws/dev/deploy.yaml"
+  build_image    = "aws/codebuild/standard:4.0"
+  build_timeout  = "10"
+
+  environment_variable = [
+    {
+      name  = "BUCKET_NAME"
+      value = "Awesome-website"
+    },
+    {
+      name  = "FRONTEND_CLOUDFRONT_ID"
+      value = var.cf_distribution_id
+    }
+  ]
+}
+
+# add policy allowing user to deploy website (s3 and cloudfront)
+resource "aws_iam_role_policy_attachment" "policy-website-to-deploy" {
+  role       = module.deploy.role
+  policy_arn = var.deploy_policy_arn
+}
+
+resource "aws_codepipeline" "codepipeline" {
+  name     = "magical-pipeline-${var.environment}"
+  role_arn = aws_iam_role.codepipeline_role.arn
+
+  artifact_store {
+    location = aws_s3_bucket.codepipeline_bucket.bucket
+    type     = "S3"
+  }
+
+  stage {
+    name = "Source"
+
+    action {
+      name             = "Source"
+      category         = "Source"
+      owner            = "ThirdParty"
+      provider         = "GitHub"
+      version          = "1"
+      output_artifacts = ["source_output"]
+
+      configuration = {
+        Owner                  = var.repo_owner
+        Repo                   = var.repo_name
+        Branch                 = var.repo_branch
+        OAuthToken             = "PLACEHOLDER"
+        "PollForSourceChanges" = "false"
+      }
+    }
+  }
+
+  stage {
+    name = "Build"
+
+    action {
+      name             = "Build"
+      category         = "Build"
+      owner            = "AWS"
+      provider         = "CodeBuild"
+      input_artifacts  = ["source_output"]
+      output_artifacts = ["build_output"]
+      version          = "1"
+
+      configuration = {
+        ProjectName = module.build.name
+      }
+    }
+  }
+
+  stage {
+    name = "Deploy-front"
+
+    action {
+      name            = "Deploy-front"
+      category        = "Build"
```

Comment:
It could be, but here i give example where its "build" category but its "Deploy"
eg.
Deploy to S3 and nothing more - is deploy because AWS supports straight copy of artifacts to S3
Deploy on kubernetes - is build that deploys on k8s because AWS don't support deploys on k8s
Deploy to S3 and cloudfront cache invalidation - is build that copies and invalidates because AWS' deploy can't invalidate cache

Reply:
As I remember it is not a label, but predefined type of step - "build" is just e.g. deploy container, and run some commands inside, "deploy" uses predefined integration to deploy artifacts somewhere (like S3, ECS, Elastic Beanstalk).
https://docs.aws.amazon.com/codepipeline/latest/userguide/reference-pipeline-structure.html

Reply:
codeDeploy has its own module for it, https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/codedeploy_deployment_config
so for that i think it would be better to create new "aws-codedeploy-step", i feel like putting them together would made everything more confusing